feat(windows): crash handler that writes minidumps to ...

ReactorScram opened the pull request:
... ProgramData/dev.firezone.client/dumps/last_crash.dmp, in debug builds only for now

For #3111.
I was way off on my estimated dump size. It's only about 500 KB.

Please double-check that the Linux exes are still stripped properly. Windows, Linux, and Mac have different ways to do embedded or non-embedded debug info, so the Windows client should still be 20 MB but with a ~62MB pdb next to it, and the Linux client should still be 15 MB.

- [ ] Check Linux artifacts weren't accidentally affected
- [ ] Make sure connlib's panic handler doesn't interfere with this
- [ ] Later on, change CI so pdb files are saved as artifacts
- [ ] Later on, enable for release builds

ReactorScram committed:
feat(windows): crash handler that writes minidumps to ProgramData/dev.firezone.client/dumps/last_crash.dmp, in debug builds only for now

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

ReactorScram committed:
fix(windows): patch profile so we don't mess up the Linux builds

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

jamilbk commented on rust/windows-client/src-tauri/Cargo.toml:
```diff
@@ -41,7 +44,7 @@ tracing-panic = "0.1.1"
 rand = "0.8.5"
 windows-implement = "0.52.0"
-known-folders = "1.1.0"
+sadness-generator = "0.5.0"
```
?? 🤣

ReactorScram replied:
It's for simulating crashes lol
It doesn't have to go to production but if anyone else needs to check my work, it lets us do segfaults and other stuff that should trigger the crash handler

jamilbk commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,162 @@
+    })
+    .context("failed to attach signal handler")?;
+
+    // On linux we can explicitly allow only the server process to inspect the
```
I'm probably missing something, but why would this be relevant inside this file for the windows client? Is it called from the linux / android codepaths?

ReactorScram replied:
Yeah good point. I just copied it from the sample code.
If we generalize the Tauri code to make a Linux GUI later it might be handy?

jamilbk commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,162 @@
+            match result {
+                Ok(mut md_bin) => {
+                    let _ = md_bin.file.flush();
+                    tracing::info!("wrote minidump to disk");
```
It might be nice to log the filepath here so our more intrepid users can inspect as they wish.

ReactorScram replied:
This happens in the crash handler process, so unfortunately these tracing logs don't actually go anywhere useful I think. I checked and even this "wrote minidump" isn't in the logs.

It might be hard for the crash handler process to append to the same connlib log file that the main process is using, too.

I was thinking of having the main process detect non-graceful shutdowns so even if the crash handler doesn't trigger, it can have a warning saying "Hey I didn't shut down gracefully last time, maybe it's a crash or maybe Task Manager killed me, check this path if you want"

conectado replied:
Are these `tracing` calls going anywhere? I don't see any subscriber.

ReactorScram replied:
They don't go anywhere. I assume it won't work to have two separate processes or subscribers writing to the same log files, but maybe I could set up a separate log dir for the crash handler process?

jamilbk commented on rust/windows-client/src-tauri/src/client/gui.rs:
```diff
@@ -38,7 +38,7 @@ pub(crate) struct Managed {
 
 // TODO: We're supposed to get this from Tauri, but I'd need to move some things around first
-const TAURI_ID: &str = "dev.firezone.client";
+pub const TAURI_ID: &str = "dev.firezone.client";
```
(nit) This is called the bundle ID in apple or package ID in android. Might be nice to have a greppable constant name to be able to update these (we should really purchase our .com)

```suggestion
pub const BUNDLE_ID: &str = "dev.firezone.client";
```

jamilbk commented on rust/windows-client/src-tauri/src/client/gui.rs:
```diff
@@ -47,10 +47,27 @@ pub(crate) fn run(params: client::GuiParams) -> Result<()> {
     let rt = tokio::runtime::Runtime::new()?;
     let _guard = rt.enter();
 
+    // If enabled, segfaults the program after 10 seconds to test crash dumps
```
I'm not the world's foremost expert in writing Rust unit tests, but would it make sense to modify the function signature for `run` to include a `segfault: bool` that we can then write a unit test for this?

Just trying to keep test codepaths out of the release codepaths (if it's not too cumbersome)

ReactorScram replied:
Yeah I'm in favor of those kind of things. We had some discussion about that before with the `inject_faults` boolean that would intentionally slow down file I/O to make sure the GUI doesn't hang. It's hard to test things like this without having some "cheat code" paths to break things on purpose.

We can also test it externally by launching the GUI as a subprocess of a unit test and then sending a kill signal to it, maybe. I'm not sure if that will trigger the crash handler on Windows. 🤔

ReactorScram replied:
Yeah it's too easy to accidentally check in an `if true` there. I added a `--crash-on-purpose` flag.

jamilbk reviewed (commented):
LGTM, just some minor suggestions and a couple questions.

I don't feel comfortable approving, so I'll defer to @conectado to approve when he's back Monday.

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

ReactorScram committed:
docs(windows): rename TAURI_ID to BUNDLE_ID and improve docs

ReactorScram committed:
feat(windows): add `--crash-on-purpose` flag

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

ReactorScram committed:
fix(windows): close the crash handler process when the main process disconnects

e.g. if it panics, we don't do crash handling for that yet, so otherwise
the crash handler server process would have leaked.

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

ReactorScram committed:
fix(windows): make sure these don't diverge

ReactorScram committed:
Merge branch 'main' into reactorscram/crash-dumps

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

ReactorScram committed:
Merge branch 'main' into reactorscram/crash-dumps

ReactorScram committed:
Merge branch 'main' into reactorscram/crash-dumps

ReactorScram committed:
Merge branch 'main' into reactorscram/crash-dumps

Signed-off-by: Reactor Scram <ReactorScram@users.noreply.github.com>

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+    attach_handler_inner()
+}
+
+#[cfg(not(debug_assertions))]
```
Why there's no handler for release?

ReactorScram replied:
I think Jamil wanted it disabled in Release mode

ReactorScram replied:
@jamilbk Is that right? Disable the crash handler in release mode? I don't remember what the conclusion was

ReactorScram replied:
Jamil said that crash dumps should be enabled in Release builds if the dumps are moved into the log dir and exported when "Export Logs" is clicked.

If I do the subprocess mitigation for #2975  there will actually be two dump files probably, one for GUI crashes and one for connlib crashes, and 3 log dirs (crash handler, GUI process, connlib process), and those 5 things will all go into the zip archive.

ReactorScram replied:
Opened #3263 to do that some other time.

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+#[cfg(not(debug_assertions))]
+pub(crate) fn attach_handler() -> anyhow::Result<crash_handler::CrashHandler> {
+    // Prevent cargo from complaining
+    if false {
```
This shouldn't be ever necessary, what kind of complain?

ReactorScram replied:
It would say that `attach_handler_inner` is dead code since the release builds wouldn't call it

conectado replied:
rather than this then we should `#[allow(dead_code)]` for for `attach_handler_inner` or maybe just don't have an inner one?

ReactorScram replied:
I pushed a commit just now to try to remove the inner function.

I wanted to keep the `target_os` marker there even though the entire client is stubbed out on non-Windows platforms, just so we know the code will compile on other platforms but not work as intended.

And since it's stubbed out, it actually should be fine if this specific module doesn't even compile on other platforms.

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+    Ok(handler)
+}
+
+fn connect() -> anyhow::Result<(Client, std::process::Child)> {
```
the `connect` naming here feels too vague

ReactorScram replied:
Yeah. Changing that to `start_server_and_connect`

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+
+const SOCKET_NAME: &str = "dev.firezone.client.crash_handler";
+
+use minidumper::{Client, Server};
```
I'd rather use the module path explicitly, `Client` and `Server` are too vague but `minidump::Client` and `minidump::Server` is easy to read.

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+    let exe = std::env::current_exe().context("unable to find our own exe path")?;
+    let mut server = None;
+
+    for _ in 0..10 {
```
why do it 10 times?

ReactorScram replied:
The sample code didn't say. Maybe in case the user tries to start a new instance of the application while the original is still writing its crash dump or something?

The sample code actually had an infinite loop, but I don't like infinite loops, so I limited it to 10.

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+    // Attempt to connect to the server
+    let (client, _server) = connect()?;
+
+    // Not sure what this does, but the sample code has it.
```
What happens if we remove it?

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+    // Not sure what this does, but the sample code has it.
+    client.send_message(1, "mistakes will be made")?;
+
+    #[allow(unsafe_code)]
```
I think it's better if we just allow unsafe code generally and we should add a `SAFETY` comment here.

ReactorScram replied:
Oops that was a copy-paste from the sample code. Removing it now

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+
+            // Send a ping to the server, this ensures that all messages that have been sent
+            // are "flushed" before the crash event is sent. This is only really useful
+            // on macos where messages and crash events are sent via different, unsynchronized,
```
we can remove this then

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+    #[allow(unsafe_code)]
+    let handler = crash_handler::CrashHandler::attach(unsafe {
+        crash_handler::make_crash_event(move |crash_context: &crash_handler::CrashContext| {
+            // Before we request the crash, send a message to the server
```
why is this needed?

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+
+    #[allow(unsafe_code)]
+    let handler = crash_handler::CrashHandler::attach(unsafe {
+        crash_handler::make_crash_event(move |crash_context: &crash_handler::CrashContext| {
```
do we need to specify the type here?

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+            // the non-crash messages are received/processed
+            client.ping().unwrap();
+
+            crash_handler::CrashEventResult::Handled(client.request_dump(crash_context).is_ok())
```
This is what sends the stack right?

ReactorScram replied:
Yes, it sends stack traces for every thread and some basis system info like CPU type and I think maybe kernel version. https://docs.rs/minidumper/latest/minidumper/struct.Client.html#method.request_dump

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+
+    // On linux we can explicitly allow only the server process to inspect the
+    // process we are monitoring (this one) for crashes
+    #[cfg(any(target_os = "linux", target_os = "android"))]
```
I'd rather add linux specific code paths when we implement the linux client and not preemptively

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+
+    for _ in 0..10 {
+        // Create the crash client first so we can error out if another instance of
+        // the Firezone client is already using this socket for crash handling.
```
could we set a random socket name so we never overlap?

ReactorScram replied:
We could, but the client is already supposed to be single-instance because it always has the same name for the wintun adapter, same file path for the device ID, etc.

conectado replied:
but if we do can we remove this loop?

ReactorScram replied:
I added a comment saying that basically I don't know why it's there so I don't know if it's okay to remove it

conectado commented on rust/windows-client/src-tauri/src/client/crash_handling.rs:
```diff
@@ -0,0 +1,172 @@
+
+    let ab = std::sync::atomic::AtomicBool::new(false);
+
+    struct Handler;
```
rather than defining it inside a function let's do a submodule?

ReactorScram replied:
I extracted it from the function, but the whole crash handling module is only 165 lines so I didn't split it yet.

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

ReactorScram committed:
windows: applying feedback for #3200

ReactorScram committed:
refactor(windows): extract Handler from function to the crash handler module

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

ReactorScram committed:
Merge branch 'main' into reactorscram/crash-dumps

ReactorScram committed:
refactor(windows): rename TAURI_ID to BUNDLE_ID (copied from #3200)

ReactorScram committed:
Merge branch 'reactorscram/rename-bundle-id' into reactorscram/crash-dumps

ReactorScram committed:
refactor(windows): move this assert after the log init, like in the crash dump PR

ReactorScram committed:
Merge branch 'reactorscram/rename-bundle-id' into reactorscram/crash-dumps

ReactorScram committed:
Merge remote-tracking branch 'origin/main' into reactorscram/crash-dumps

conectado reviewed (approved):
LGTM! left a few responses but nothing blocking

ReactorScram committed:
refactor(windows): try to simplify attach_handler

ReactorScram committed:
docs(windows): add comment
